docs/DevGuide/WritingTests.md

```diff
@@ -4,12 +4,167 @@ See LICENSE.txt for details.
+this case `DataStructures`. The `SPECTRE_TEST_CASE` macro should be treated as a
+function, which means that it should be followed by `{ /* test code */ }`. For
+example,
+\snippet Test_Tensor.cpp example_spectre_test_case
+From within a `SPECTRE_TEST_CASE` you are able to do all the things you would
+normally do in a C++ function, including calling other functions, setting
+variables, using lambdas, etc.
+
+The `CHECK` macro in the above example is provided by
+[Catch2](https://github.com/catchorg/Catch2) and is used to check conditions. We
+also provide the `CHECK_ITERABLE_APPROX` macro which checks if two `double`s or
+two iterable containers of `double`s are approximately
+equal. `CHECK_ITERABLE_APPROX` is especially useful for comparing `Tensor`s,
+`DataVector`s, and `Tensor<DataVector>`s since it will iterate over nested
+containers as well.
+
+\warning Catch's `CHECK` statement only prints numbers out to approximately 10
+digits at most, so you should generally prefer `CHECK_ITERABLE_APPROX` for
+checking double precision numbers, unless you want to check that two numbers are
+bitwise identical.
+
+All unit tests must finish within a few seconds, the hard limit is 5, but having
+unit tests that long is strongly discouraged. They should typically complete in
+less than half a second. Tests that are longer are often no longer testing a
+small enough unit of code and should either be split into several unit tests or
+moved to an integration test.
+
+#### Discovering New and Renamed Tests
+
+When you add a new test to a source file or rename an existing test the change
+needs to be discovered by the testing infrastructure. This is done by building
+the target `rebuild_cache`, e.g. by running `make rebuild_cache`.
+
+#### Testing Pointwise Functions
+
+Pointwise functions should generally be tested in two different ways. The first
+is by taking input from an analytic solution and checking that the computed
+result is correct. The second is to use the random number generation
+comparison with python infrastructure. In this approach the C++ function being
+tested is re-implemented in python and the results are compared. If the function
+does sums over tensor indices then the einsum package should be used in python
+to provide an alternative implementation of the loop structure. The python
+implementations should be in a file name `TestFunctions.py` in the same
+directory as the `Test_*.cpp` source file and have the same name as the C++
+function being tested. It is possible to test C++ functions that return by value
+and ones that return by `gsl::not_null`. In the latter case, since it is
+possible to return multiple values, one python function taking all
+non-`gsl::not_null` arguments must be supplied for each `gsl::not_null` argument
+to the C++. To perform the test the `pypp::check_with_random_values()` function
+must be called. For example, the following checks various C++ functions by
+calling into `pypp`:
+
+snippet Test_Pypp.cpp HALP
```

Comment:
Should there be an actual snippet here?

Reply:
Yes, was waiting for the python PR to get merged :)